text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,183 @@
+
+#### 2.3 Resolved Timestamp
+
+**TiKV-CDC** using *Resolved Timestamp* (or `resolved-ts`) to indicate the largest transaction timestamp of the replication task, to help downstream avoid getting partial data of a transaction. Downstream streaming system can also use `resolved-ts` as [*watermark*](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/).
+
+As no transaction support, in RawKV scenario we don't need scanning locks in *lock_cf*. But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).
+
+### 3. Deletion
+
+To make deletions be captured as change data, we turn physical deletion to logical deletion, i.e., set a *deleted* flag, other than physically delete the entry.
+
+The *garbage collection* of deleted data is implemented by setting TTL to 25 hours after, and *GC-ed* by compaction filter. At the same time, start timestamp (i.e. `start-ts`) of **TiKV-CDC** replication task should not be earlier than 24 hours before.
+
+### 4. Encoding
+
+Timestamp & deleted flag are encoded as following:
+
+```
+r{keyspace id}{user key}{MCE Padding}{^timestamp:u64}: {user value}{expire ts}{meta flags}
+```
+
+*(Deleted flag is a bit of meta flags in value)*
+
+* Encoding based on TiKV API version [V2](https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md). *(As there is no flag or meta fields, it is not possible to encode the two necessary fields for CDC in older API version.)*
+* Timestamp encoded into key will be beneficial to the following, comparing with encoded into value:
+  * More effective catch up, as we can get scope of change data without read & decode all values.
+  * [PiTR](https://en.wikipedia.org/wiki/Point-in-time_recovery) is available.
+  * Less code complexity, as data structure of RawKV & TxnKV are the same.
+* And with the following acceptable loss:
+  * More read latency in tens of microseconds.
+  * A little more write overhead caused by [MCE Padding](https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format).
+  * An extra Garbage Collection is required (used with TxnKV).
+  * Possible more space occupied (related to update pattern).
+
+### 5. Incremental Scan
+
+On replication task resumed after paused, we scan *KVDB* to get the change data during pause by extracting timestamp encoded in key. The scan will be heavy as we actually scan all the keys, so we should avoid scanning as possible, and limit scan rate.
+
+*Incremental Scan in SQL / TxnKV scenario scans write_cf to get change data scope, which is much cheaper. So an alternative to reduce cost for RawKV is writing timestamp to write_cf other than encoded in key, but will introduce an additional write for each put request. We tend to the lower delay of writing now.*
```

Comment:
what's the solution for writing `timestamp`?

Reply:
It would be: in addition to `{key}:{value}` in default_cf, write an entry `{key}:{timestamp}` to write_cf, to indicated the latest change timestamp of the key.